### ISSUE TITLE
Update LC Permit expiration batch email templates and 48 hour batch timer limit (SS) (In Discussion)

### ISSUE THREAD

**Issue by TracyLinder** [cityofaustin/atd-data-tech#16106]
**Acceptance Criteria** 

- The 24 hour limit logic for checking expired LC licenses set up in #10849 should be removed from the batch process.
- The email template needs to be updated to the following:
### Expiration email template

Subject: ROW License and Permit Expiration Notice

Dear ROW License Holder,

This email is to advise you that your Right of Way Contractor License [2022-12345678 LC] expired on [LC folder Expiry date].

All Driveway/Sidewalk and Excavation permits associated with your Right of Way Contractor License have expired on [LC folder Expiry date] until the license is renewed. To renew your Right of Way Contractor License, look for the "How is a Right of Way Contractor License renewed?" Section at this URL - https://austintexas.gov/page/right-way-contractors

The following list includes the DS/EX Permits that have expired. To view a full list of expired permits, please visit https://abc.austintexas.gov/web/permit/folder-select?reset=true

    [Folder Number of DS or EX folder, for example, 2022-012345 DS] - [Folder Name]
        2023-12345678-DS - New Driveway & Sidewalk

To reactivate your license, please view the documentation requirements on the Right of Way Contractor webpage {https://www.austintexas.gov/page/right-way-contractors}, and then submit your updated documentation to our SmartSheet {https://app.smartsheet.com/b/form/cad1e53c88224399979f40df80ef6270}.

### Re-activation email template

Subject: ROW License and Permit Renewal Notice

Dear ROW License Holder,

This email is to advise you that your Right of Way Contractor License [2022-12345678 LC] has been reactivated.

All Driveway/Sidewalk and Excavation permits associated with your Right of Way Contractor License have been reactivated. Please note that any permits that are past their expiry date will not be reactivated.

The following list includes the DS/EX Permits that have been Reactivated. To view a full list of reactivated permits, please visit  https://abc.austintexas.gov/web/permit/folder-select?reset=true

    [Folder Number of DS or EX folder, for example, 2022-012345 DS] - [Folder Name]
        Example: 2023-12345678 DS - New Driveway & Sidewalk

- The number of permits listed should be *TBD* (Need Raj and Subha to confirm)
- If there are no DS/EX permits to list, system must display 'N/A' where the list of permits would have displayed. (from #15763)

Notes from original ticket description:
Determine why folder 13045443 wasn't expired-license when the related license expired (12753417)

The license for Captex Construction, LLC expired on 2/22/24 however it's child permit (13045443 ) did not update to expired permit, it is still active.

**Comment by TracyLinder** [comment]
There also aren't any EX permits with the status expired-license, indicating an issue/bug.

**Comment by SubhaShroff** [comment]
@TracyLinder  @Nadin-Nader @vschandramouli 
Batch - Expire EX DS with LC(Job RSN-203):
if we don't put expired date/ or the date it was expired check  in the Cursor to get the expired LC and ROW contractor, everyday the batch will hit all the expired LCs irrespective of the expirydate/stampdate. 
So, my question would be - are there any scenarios that DS/EX can get reactivated (other than renewing the LC with Renew Process), before even LC is renewed/reactivated?
1- If any such scenarios exist, while LC is still expired DS/EX are activated, but the batch will keep updating the DS/EX 'Active' status to 'Expired-License' status.
2- As you suggested to check the update date(stampdate) instead of expirydate, I am not sure if with the status update the stampdate is also updated. To know that, I would need to know what are the processes/batches  or any other triggerring points that expire LC. So that I make sure those are all updating the stamp date also with status update.

**Comment by TracyLinder** [comment]
### DTS-AMANDA Conversation 3/5/24

- We know that the where clause with expiration the last 24 hours is causing this issue.
- We know that we added that clause, but we can't remember why
- We are going to remove this and test it again, and go from there.

**Comment by SubhaShroff** [comment]
@TracyLinder @Nadin-Nader 
Expirydate WHERE clause commented and the commit is sent to DEV - SCTASK0401844.
Thanks.

**Comment by Nadin-Nader** [comment]
@SubhaShroff this has been sent to the DevOps CM Team to be deployed to DEV under **SCTASK0402207**

**Comment by SubhaShroff** [comment]
@Nadin-Nader  The above DEV deployment I tested but batch job is throwing overflow error - "PL/SQL: numeric or value error: character string buffer too small ORA-06512:"
As we discussed today (3/7/2024) about this issue, we can only test the expiry date condition commented after fixing the above issue addressed.

I checked with the code if it will do with permit number format(2024-021365 EX) only, it still overflows.. I tried only for RSNs listing, that too fails for these Right of Contractors permits with the given number of active permits(EX, DS) they have. So, we only can make it work by limiting the display of these permits in the email, let be RSNs alone. Below data are on PROD only a few for example - contractors with active EX, DS
 
![image](https://github.com/cityofaustin/atd-data-tech/assets/51680715/33c9bcf7-b68c-43d1-b7e0-1271af012dea)

NOTE:
This is working fine as of now since its not hitting all the LCs, especially with those having large number of permits. 

Thanks.

**Comment by TracyLinder** [comment]
### Solution is to have 2 notification templates, one that includes a list of permits and one that contains a link to the portal instead. 

does the character limit error message happen?
Y - Send email with link to portal `we need email content for this`
N - Send email with the permits listed



Emails that need to be created:
1. Expiration Notification with portal link
2. Renewal Notification with portal link (renew process on LC)


**Next Steps:**
1. Turn the batch job(s) off.
2. Creating email content
3. Subha create the new emails 
4. Subha update the batch job with email logic based on the character limit

**Comment by TracyLinder** [comment]
@Vaishnav1397 can you please work with Kyle to create email templates that point the user to the portal to find their permits. Please also create AC for the fix.

**Comment by vschandramouli** [comment]
Moving this back to Needs Scoping to work with Kyle.

**Comment by Nadin-Nader** [comment]
Incident to Pause Batch Jobs on PROD **INC0374203**

**Comment by TracyLinder** [comment]
We will need to document the use cases that require UAT.
1. Folder expired because it expired
2. Folder expired b/c insurance expired
3. 2. Folder expired b/c bond expired

**Comment by TracyLinder** [comment]
### Note

The causes of this bug were identified as 
1. character limit issues with the automated emails.
2. A 24 hour limit in the query where clause

Resolutions:
1. Create new email templates for the scenario where the list gets too long to prevent the character limit issue
2. Remove the 24 hour limit in the query

**Comment by vschandramouli** [comment]
Notes from scoping meeting with Kyle and Raj: 4/2/24

- Raj recommends showing a set number of expiring permits (say 5 to 10), and still adding a link to the portal so they can view the full list. 
- Subha and Raj can figure out what the threshold is before we start seeing character limit errors, and we can confirm with ROW teams to set it. 
- Vish will add AC from #15763 as part of this card. 

Subject: ROW License and Permit Expiration Notice

Dear ROW License Holder,

This email is to advise you that your Right of Way Contractor License [2022-12345678 LC] expired on [LC folder Expiry date].

All Driveway/Sidewalk and Excavation permits associated with your Right of Way Contractor License have expired on [LC folder Expiry date] until the license is renewed. A list of those permits is provided below. To renew your Right of Way Contractor License, look for the "How is a Right of Way Contractor License renewed?" Section at this URL - https://austintexas.gov/page/right-way-contractors

The following list includes the DS/EX Permits that have expired. To view a full list of expired permits, please visit **add link to the AB+C Portal here**.

    [Folder Number of DS or EX folder, for example, 2022-012345 DS] - [Folder Name]
        2023-12345678-DS - New Driveway & Sidewalk

To reactivate your license, please view the documentation requirements on the Right of Way Contractor webpage {https://www.austintexas.gov/page/right-way-contractors}, and then submit your updated documentation to our SmartSheet {https://app.smartsheet.com/b/form/cad1e53c88224399979f40df80ef6270}.

**Comment by Nadin-Nader** [comment]
This has been sent to the DEVOps CM Team to be deployed to DEV under **SCTASK0412009**

**Comment by Nadin-Nader** [comment]
@SubhaShroff this has been sent to the DevOps CM Team to be deployed in TEST under **SCTASK0413275**

**Comment by SubhaShroff** [comment]
@Nadin-Nader SCTASK0413275 is closed but looks like one of the commits is not deployed in TEST - fd0f598f26274f006dd96d7886b2807f1dc89b25. Messaged Monty about it. Will retest it once its deployed properly. Thank you.

**Comment by SubhaShroff** [comment]
@Nadin-Nader  I tested this after Monty fixed it.  
Right-of-Way Contractor - 5747087
LC RSN - 12623304
The expiration batch email is attached. This particular Right of Contractor had 32 EX/DS active permits. The email listed those permits only as much as it could accommodate. 
All those active permits' status got updated to Expired-License.


![image](https://github.com/cityofaustin/atd-data-tech/assets/51680715/c37d2d1c-2a44-43e7-8b38-1d9561dbb5d8)

Yet to test Renew .process...............

**Comment by SubhaShroff** [comment]
@Nadin-Nader  Please create a DEV deployment request. Thank you.
Fixed the renewal email content and the commit -8bd726c7875679d3bc8069036545729356bfe717 is sent for DEV deployment.
[16106 - 8bd726c7875679d3bc8069036545729356bfe717 - DEV Deployment Plan.docx](https://github.com/cityofaustin/atd-data-tech/files/15113369/16106.-.8bd726c7875679d3bc8069036545729356bfe717.-.DEV.Deployment.Plan.docx)

**Comment by Nadin-Nader** [comment]
this has been sent to the DevOps CM Team to be deployed in DEV **SCTASK0415921**

**Comment by Nadin-Nader** [comment]
This has been sent to the DevOps CM Team to be deployed to TEST under **SCTASK0416157**

**Comment by SubhaShroff** [comment]
@Nadin-Nader This is ready for UAT on TEST. Thank you.

**Comment by SubhaShroff** [comment]
@Nadin-Nader @vschandramouli @KyleErich @TracyLinder 
After testing this ticket on TEST there are few points I noted that seek your view:
### 1. Turning the Batch job on is triggering it for all the previous days that it was not on. 
Suggested Solution:
To avoid it, on test environment we can run Batch job manually, which will fire only once. On PROD, I think we will delete the existing Batch and add it again as fresh.

### 2. The Batch job (Update DS/EX with expired LC), would send email to the Right of Way Contractor daily until the LC is Activated or atleast the status is updated to something that is not EXPIRED.
This is the reason Tracy had suggested to add an expiry date check and do the update/send emails to only those 
LC licenses that expired previous day. That should restrict to send the emails everyday. 
But This ticket has some examples that would be an exception for the date check.
- One, LC which doesn't have an expiry date on it but expired- If the expiry date check is added, these LCs are never addressed by the batch job.
- Two, the old expired LCs with past exipry date (other than previous day).
- If we still consider only the previous day expired LCs, to address the old LCs we can run a script, for once. 
That way this date check solution can be made workable and not send repeated emails to ROW Contractors for the expired LC until it is renewed.

But, there will be a question about the LCs that are expiring without expiry date on it, after the script is executed. I think, adding expiry date to the LC along with the expired status update would solve that too. (Need to know when it expires but not update the expiry date).

Or Run the batch weekly once to reduce the email frequency. This will also, update the DS/EX permits weekly once, to Expired-License.

Please let me know your thoughts/solutions that works. Thank you.

**Comment by KyleErich** [comment]
1. I agree with the suggested solution.
2. I'd like to avoid the emails going out daily or even weekly. To me that will overburden our customers and result in consternation/complaints. In regards to the scenario where the folder is Expired but there's no Expiry Date, I don't think that is a common scenario. Any LC that has been issued will have an expiry date -- presumably most of the ones without an expiry date were never issued in the first place and whose last status before Expired was Pending. It's okay to me if these cases don't receive an email.

We can discuss in greater detail during the Friday afternoon meeting.

**Comment by TracyLinder** [comment]
### DTS-AMANDA Refinement Conversation 5/7/24

**Outcome:** We suggest to fix the insurance expiration date issue and add a 'expired yesterday' clause to the batch job.


### Problem 1: The batch job is emailing LC Right-of-Way Contractor every day
- Solution would be insert a folder comment to record the notification email has been sent. The batch would be updated so as not to send a notification email for those that have already gotten an email. 
- This introduces a new problem of the batch job only working one time; the 1st time. So the folder comment would need to be deleted after the LC is reactivated/renewed. 




### Problem 2: When insurance expires the LC, the expiration date remains in the future so the batch does not pick it up. 
- Fixing this would resolve the 1st problem. 
- Solution would be to update the expiration procedure to input today's date as the expiration date.
- We'd need to be sure the expiration date gets fixed when the folder is reactivated. 
- Be sure the Expired-License batch job only picks up licenses with yesterday's expiration date.

**Comment by vschandramouli** [comment]
Notes from meeting with Subha and Raj 

- Raj mentions there is no fall back if this batch fails one day. Since batch checks only 24 hours, if the batch fails one day then it wont get caught in the future 
- Subha mentioned it may be helpful to split solution for problem 1 and problem 2 into 2 separate tickets since they are two different batch jobs. 
- Raj mentioned potentially using a Read-only folder info with radio buttons (like Expired email sent? Yes/No). Read only means users will not be able to edit this. When system sends email, update folder info to Yes. When folder is activated (which status is considered activated? Just active? Something else?), update folder info to No. 
- This info may also serve as a record of emails sent to the applicant. 
- We will need to add the info first, re-default all folders and then add/run the batch.
- Vish will add this to Friday meeting agenda.

**Comment by vschandramouli** [comment]
Meeting with Kim, Paloma, Kyle and team - 

- It would be okay to extend the yesterday clause to 48 hours, it would be okay to send 2 emails to ROW Contractor. 
- This will reduce complexity of adding info field. 
- We will proceed with this approach. Vish will update AC.

**Comment by vschandramouli** [comment]
Vish will create a separate card for insurance work (problem 2) and scope it out with ROW team. This may be related to #16314

**Comment by TracyLinder** [comment]
### Update - On Hold
- We have decided to update the insurance expiration scenario prior to updating the batch job with the 48 hour clause.

**Comment by SubhaShroff** [comment]
@Nadin-Nader Please create a DEV deployment request. Thank you.
[16106 - 56985c629df0348434b63f75c8bdee7b3ec069a7 - DEV Deployment Plan.docx](https://github.com/user-attachments/files/16183716/16106.-.56985c629df0348434b63f75c8bdee7b3ec069a7.-.DEV.Deployment.Plan.docx)

**Comment by Nadin-Nader** [comment]
This has been sent to the DevOps CM Team to be deployed to DEV under **SCTASK0439363**

**Comment by Nadin-Nader** [comment]
Status:
Waiting on TEST UAT for #16415. If #16415 is approved on TEST, this one can progress

**Comment by Nadin-Nader** [comment]
#16415 has passed UAT on TEST and is going to PROD. This has been sent to the CM DevOps Team for deployment to TEST under **SCTASK0441813**

**Comment by SubhaShroff** [comment]
Set the data to TEST:
PeopleRSN-5597614	LC RSN-11766283

13150693	DS	2024-10-26	50010
13193141	EX	2024-07-26	50010
12819803	DS	2025-01-22	50010

The following batches will trigger tonight after midnight.
196 - Expire EX DS with LC
179 - Expire EX Folders
200 - Expire DS with Expire License Status

Will verify the data on Monday.

**Comment by Nadin-Nader** [comment]
The new commit has been sent to the DevOps CM Team to be deployed to DEV under **SCTASK0443101**

**Comment by Nadin-Nader** [comment]
@SubhaShroff This has been deployed to DEV please UAT

**Comment by SubhaShroff** [comment]
DEV Batches turned on : 
198 - Expire EX DS with LC
204 - Expire DS with Expire License Status

**Comment by SubhaShroff** [comment]
@Nadin-Nader @TracyLinder  Expire EX folders batch job(already existing not part of 10849 ticket). 
This one expires EX permits with expirydate of the previous day. So, old EX permits if there are any with past expiry date other than previous day. **Should I remove the previous day expiry date condition for this Batch Job?**

NOTE: And also this batch check not only for 'Expired-License' status of EX, but also other status which are NOT any of the following status  -  XXXX Expired, Final, Canceled, Closed, Void, Aborted  XXXX

**Comment by Nadin-Nader** [comment]
Another DEV commit has been sent to the DevOps CM TEAM to be deployed to DEV under **SCTASK0444016**

**Comment by Nadin-Nader** [comment]
@SubhaShroff this has still not been deployed as of 10:07am 8/2/24

**Comment by Nadin-Nader** [comment]
@SubhaShroff This has been deployed to DEV please UAT

**Comment by SubhaShroff** [comment]
@Nadin-Nader Thank you.
Setting the data to test the following batch jobs:
198 - Expire EX DS with LC
204 - Expire DS with Expire License Status
154 - Expire EX folders

Ppl           LCrsn               status       expirydate
5723093	12482138	50030	2024-07-31
5418121	11113820	50030	2024-08-03
5810048	13077576	50030	2024-08-01
5815138	13106720	50030	2024-08-02


There are many EX/DS. Few are listed to verify on Monday.
11109909	DS
12616443	EX
12596371	EX
13181667	EX
13166805	EX
13193483	EX
13187863	EX
11109909	DS
13181843	DS
13112204	DS
12886853	DS
11109909	DS
13181843	DS
13112204	DS
12886853	DS
11109909	DS

**Comment by Nadin-Nader** [comment]
This has been sent to the DevOps CM Team to be deployed on DEV under **SCTASK0445323**

**Comment by Nadin-Nader** [comment]
This has been deployed to DEV. Subha is checking on it but also has to discuss the email portion of it @ the 2pm meeting with ROW

**Comment by Nadin-Nader** [comment]
This has been sent to the DevOps CM Team to be deployed to DEV under**SCTASK0447019**

**Comment by Nadin-Nader** [comment]
This has been sent to the DevOps CM Team to be deployed to TEST under **SCTASK0448365**

**Comment by SubhaShroff** [comment]
@Nadin-Nader I am testing this on TEST. Tested day1, today. Will need to test it on day 2 as well, checking the emails tomorrow.

**Comment by SubhaShroff** [comment]
@Nadin-Nader Since DS expire batch was not working I couldn't test it completely on DEV. Its tested on TEST. 2nd day email is not good enough. And did another commit for the fix. - SCTASK0450580

**Comment by Nadin-Nader** [comment]
This has been sent to the DevOps CM Team to be deployed in TEST under **SCTASK0450612**

**Comment by SubhaShroff** [comment]
@Nadin-Nader  @KyleErich  This is done on TEST. I have tested it, ready for UAT on TEST.

### NOTE: I need to turn the jobs on whenever Kyle is going to UAT to avoid emails flooding in others inboxes.

Thank you.

**Comment by KyleErich** [comment]
@Nadin-Nader @SubhaShroff I'm ready to UAT this one. This serves as my notification to turn the jobs on.

**Comment by TracyLinder** [comment]
Pending UAT by ROW

**Comment by TracyLinder** [comment]
ROW was informed that this is at risk of being overwritten in the test environment. This is a hard feature to test, especially alone. Kyle/Vish/Russ will work on this together next week.

**Comment by Nadin-Nader** [comment]
Vish will be creating a meeting with Kyle and Subha to UAT

**Comment by vschandramouli** [comment]
@Nadin-Nader @SubhaShroff UAT approved by Kyle and Russell. Ready for Prod Deployment.

**Comment by SubhaShroff** [comment]
Thank you.
@Nadin-Nader  I see, there is change related to 'Renew License' LC process in #16314. Will coordinate with him for deployments.

**Comment by SubhaShroff** [comment]
@Nadin-Nader  SCTASK0468029 - TEST promotion request is created. Thanks.

**Comment by SubhaShroff** [comment]
@Nadin-Nader Renewing LC process, sends email. Noticed email was not listing the activated permits. Now it does. Please UAT this part on TEST. Thank you.

**Comment by KyleErich** [comment]
@Nadin-Nader @SubhaShroff UAT on Test approved.

**Comment by Nadin-Nader** [comment]
This has been sent to the DevOps CM Team to be deployed to PROD under **SCTASK0469397**

**Comment by TracyLinder** [comment]
### Update
- When we promoted this to prod it had a bug that expired all EX folders in the system. (See #19877)
- This job is currently turned off and devs have re-written it; 
- LIT will be working on documenting the entire logic from end to end. 
- Then devs will discuss with ROW the logic
- Then devs will update code as needed and peer review it internally.
- Then we can discuss turning it back on. _**Can we get ready for this discussion by 12/6?**_
- Also discuss if we need to run a script to update Active folders to Expired License status

**Comment by Nadin-Nader** [comment]
This has been sent to the DevOps CM Team to be deployed to DEV and TEST again with new changes under **SCTASK0474371**

**Comment by SubhaShroff** [comment]
@Nadin-Nader I verified the deployment on TEST batches. There are compilation errors on COA_BATCH_PROCESS Package with the deployment. I need to check with Devops Team. I will be reaching Hari in the morning, as he was the one deployed it.
Thank you.

**Comment by TracyLinder** [comment]
### DTS-AMANDA Conversation about the Bond/Insurance expiration issue

- We don't think updating the LC to "expired-license" is going to work, b/c the batch job is only designed to pickup Expired LC folders to trigger the batch. 
- Tracy wants to create an epic for this functionality.
- We need to regroup and try again on a proper solution.

**Comment by vschandramouli** [comment]
Updated batch documentation is below:

[Updated Batch documentation.docx](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBLzRsQ0E9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--f5a6c38ee9a3ab0485b9244dbb8ff32f1313358e/Updated%20Batch%20documentation.docx)

**Comment by vschandramouli** [comment]
Batch solution updated in tabular format for team review. Attached below. [Updated Batch documentation 12-20.docx](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBNVU2Q0E9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--4d16bbbc2ba100bc89bef8592e4257dd044eac1d/Updated%20Batch%20documentation%2012-20.docx)

**Comment by SubhaShroff** [comment]
@Nadin-Nader  @vschandramouli @TracyLinder  Is this still ready for development?
Thanks.

**Comment by TracyLinder** [comment]
> @Nadin-Nader @vschandramouli @TracyLinder Is this still ready for development? Thanks.

Let's discuss this at our team meeting tomorrow.

**Comment by TracyLinder** [comment]
We're going to close this and begin working on the new solution Build new (combined) LC Expiration batch job #20380
